File: examples/javascript/metadata_client.js
/**
 * 图像元数据检测与取证分析系统客户端
 */
class ImageForensicsClient {
  /**
   * 初始化客户端
   * @param {string} baseUrl - API服务的基础URL
   */
  constructor(baseUrl = 'http://localhost:8080') {
    this.baseUrl = baseUrl;
  }

  /**
   * 检查API服务健康状态
   * @returns {Promise<Object>} 包含健康状态的对象
   */
  async checkHealth() {
    try {
      const response = await fetch(`${this.baseUrl}/health`);
      return await response.json();
    } catch (error) {
      return { status: 'error', message: error.message };
    }
  }

  /**
   * 提取图像元数据
   * @param {File} imageFile - 图像文件对象
   * @returns {Promise<Object>} 包含元数据的对象
   */
  async extractMetadata(imageFile) {
    try {
      const formData = new FormData();
      formData.append('image', imageFile);

      const response = await fetch(`${this.baseUrl}/metadata`, {
        method: 'POST',
        body: formData
      });

      return await response.json();
    } catch (error) {
      return { status: 'error', message: error.message };
    }
  }

  /**
   * 批量提取多个图像的元数据
   * @param {Array<File>} imageFiles - 图像文件对象数组
   * @returns {Promise<Object>} 包含所有图像元数据的对象
   */
  async batchExtractMetadata(imageFiles) {
    try {
      const formData = new FormData();
      
      imageFiles.forEach(file => {
        formData.append('images[]', file);
      });

      const response = await fetch(`${this.baseUrl}/metadata/batch`, {
        method: 'POST',
        body: formData
      });

      return await response.json();
    } catch (error) {
      return { status: 'error', message: error.message };
    }
  }

  /**
   * 分析图像是否被篡改
   * @param {File} imageFile - 图像文件对象
   * @returns {Promise<Object>} 包含取证分析结果的对象
   */
  async analyzeForensics(imageFile) {
    try {
      const formData = new FormData();
      formData.append('image', imageFile);

      const response = await fetch(`${this.baseUrl}/forensics`, {
        method: 'POST',
        body: formData
      });

      return await response.json();
    } catch (error) {
      return { status: 'error', message: error.message };
    }
  }
}

// 在浏览器环境中使用
if (typeof window !== 'undefined') {
  window.ImageForensicsClient = ImageForensicsClient;
}

// 在Node.js环境中使用
if (typeof module !== 'undefined' && module.exports) {
  module.exports = ImageForensicsClient;
}

/**
 * 使用示例（浏览器环境）
 */
async function browserExample() {
  // 创建客户端实例
  const client = new ImageForensicsClient();
  
  // 检查API健康状态
  const healthStatus = await client.checkHealth();
  console.log('API健康状态:', healthStatus);
  
  // 获取文件输入元素
  const fileInput = document.getElementById('imageFile');
  if (fileInput && fileInput.files.length > 0) {
    const imageFile = fileInput.files[0];
    
    // 提取元数据
    console.log(`从 ${imageFile.name} 提取元数据...`);
    const metadata = await client.extractMetadata(imageFile);
    console.log('元数据结果:', metadata);
    
    // 取证分析
    console.log(`对 ${imageFile.name} 进行取证分析...`);
    const forensics = await client.analyzeForensics(imageFile);
    console.log('取证分析结果:', forensics);
    
    // 判断图像是否被篡改
    if (forensics.status === 'success') {
      const isTampered = forensics.forensics?.is_tampered;
      if (isTampered) {
        console.warn('警告: 图像可能已被篡改!');
        const indicators = forensics.forensics?.tampering_indicators || [];
        indicators.forEach((indicator, index) => {
          console.warn(`篡改指标 ${index + 1}: ${indicator.description}`);
        });
      } else {
        console.log('图像未检测到篡改痕迹');
      }
    }
  }
} 

File: examples/python/metadata_client.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import requests
import json
import os
import argparse
from typing import Dict, List, Optional, Union, Any


class ImageForensicsClient:
    """图像元数据检测与取证分析系统客户端"""

    def __init__(self, base_url: str = "http://localhost:8080"):
        """初始化客户端

        Args:
            base_url: API服务的基础URL
        """
        self.base_url = base_url

    def check_health(self) -> Dict[str, Any]:
        """检查API服务健康状态

        Returns:
            Dict: 包含健康状态的字典
        """
        try:
            response = requests.get(f"{self.base_url}/health")
            return response.json()
        except Exception as e:
            return {"status": "error", "message": str(e)}

    def extract_metadata(self, image_path: str) -> Dict[str, Any]:
        """提取图像元数据

        Args:
            image_path: 图像文件路径

        Returns:
            Dict: 包含元数据的字典
        """
        if not os.path.exists(image_path):
            return {"status": "error", "message": f"文件不存在: {image_path}"}

        try:
            with open(image_path, 'rb') as image_file:
                files = {'image': image_file}
                response = requests.post(f"{self.base_url}/metadata", files=files)
            return response.json()
        except Exception as e:
            return {"status": "error", "message": str(e)}

    def batch_extract_metadata(self, image_paths: List[str]) -> Dict[str, Any]:
        """批量提取多个图像的元数据

        Args:
            image_paths: 图像文件路径列表

        Returns:
            Dict: 包含所有图像元数据的字典
        """
        files = []
        try:
            for path in image_paths:
                if not os.path.exists(path):
                    return {"status": "error", "message": f"文件不存在: {path}"}
                files.append(('images[]', open(path, 'rb')))

            response = requests.post(f"{self.base_url}/metadata/batch", files=files)
            
            # 关闭所有打开的文件
            for _, file_obj in files:
                file_obj.close()
                
            return response.json()
        except Exception as e:
            # 确保关闭所有已打开的文件
            for _, file_obj in files:
                try:
                    file_obj.close()
                except:
                    pass
            return {"status": "error", "message": str(e)}

    def analyze_forensics(self, image_path: str) -> Dict[str, Any]:
        """分析图像是否被篡改

        Args:
            image_path: 图像文件路径

        Returns:
            Dict: 包含取证分析结果的字典
        """
        if not os.path.exists(image_path):
            return {"status": "error", "message": f"文件不存在: {image_path}"}

        try:
            with open(image_path, 'rb') as image_file:
                files = {'image': image_file}
                response = requests.post(f"{self.base_url}/forensics", files=files)
            return response.json()
        except Exception as e:
            return {"status": "error", "message": str(e)}


def print_json(data: Dict[str, Any]) -> None:
    """美化打印JSON数据

    Args:
        data: 要打印的字典数据
    """
    print(json.dumps(data, indent=2, ensure_ascii=False))


def main():
    """主函数"""
    parser = argparse.ArgumentParser(description='图像元数据检测与取证分析客户端')
    parser.add_argument('--url', default='http://localhost:8080', help='API服务的URL')
    parser.add_argument('--action', choices=['health', 'metadata', 'batch', 'forensics'], 
                        required=True, help='要执行的操作')
    parser.add_argument('--image', help='图像文件路径')
    parser.add_argument('--images', nargs='+', help='多个图像文件路径（用于批量处理）')
    
    args = parser.parse_args()
    
    client = ImageForensicsClient(args.url)
    
    if args.action == 'health':
        print("检查API健康状态...")
        result = client.check_health()
        print_json(result)
        
    elif args.action == 'metadata':
        if not args.image:
            print("错误: 提取元数据需要指定--image参数")
            return
        print(f"从 {args.image} 提取元数据...")
        result = client.extract_metadata(args.image)
        print_json(result)
        
    elif args.action == 'batch':
        if not args.images:
            print("错误: 批量提取元数据需要指定--images参数")
            return
        print(f"批量提取 {len(args.images)} 个图像的元数据...")
        result = client.batch_extract_metadata(args.images)
        print_json(result)
        
    elif args.action == 'forensics':
        if not args.image:
            print("错误: 取证分析需要指定--image参数")
            return
        print(f"对 {args.image} 进行取证分析...")
        result = client.analyze_forensics(args.image)
        print_json(result)
        
        # 判断图像是否被篡改
        if result.get("status") == "success":
            is_tampered = result.get("forensics", {}).get("is_tampered", False)
            if is_tampered:
                print("\n警告: 图像可能已被篡改!")
                indicators = result.get("forensics", {}).get("tampering_indicators", [])
                for i, indicator in enumerate(indicators, 1):
                    print(f"篡改指标 {i}: {indicator.get('description')}")
            else:
                print("\n图像未检测到篡改痕迹")


if __name__ == "__main__":
    main() 